MESIFCacheSet.cpp

```diff
@@ -0,0 +1,191 @@
+                }
+
+                bus->setHasForwardState(true);
+            }
+        }
+        case CacheLineState::SHARED:
+        case CacheLineState::EXCLUSIVE:
+        case CacheLineState::MODIFIED:
+        case CacheLineState::FORWARD:
+            break;
+        default:
+            throw std::logic_error("Invalid cache state for MESIF");
+    }
+
+    if (node->state == CacheLineState::EXCLUSIVE
+            || node->state == CacheLineState::MODIFIED) {
+        logger->incrementPrivateDataAccess();
+    } else {
+        logger->incrementPublicDataAccess();
+    }
+
+    return numCycles;
+}
+
+int MESIFCacheSet::write(
+    int threadID,
+    uint32_t tag,
+    std::shared_ptr<Bus> bus,
+    std::shared_ptr<Logger> logger
+) {
+    int numCycles = CacheSet::write(threadID, tag, bus, logger);
+    std::shared_ptr<CacheLineNode> node = this->cacheSet[tag];
+
+    switch(node->state) {
+        case CacheLineState::INVALID: {
+            bool isExclusive = bus->busReadExclusiveAndCheckIsExclusive(
+                this->getBlockIdx(tag),
+                threadID
+            );
+
+            if(isExclusive) {
+                numCycles += MEMORY_FETCH_COST;
+
+                logger->addExecutionCycles(MEMORY_FETCH_COST);
+                logger->addIdleCycles(MEMORY_FETCH_COST);
+            } else {
+                numCycles += this->numCyclesToSendBlock;
+
+                logger->incrementBusTraffic();
+                logger->addExecutionCycles(this->numCyclesToSendBlock);
+                logger->addIdleCycles(this->numCyclesToSendBlock);
+            }
+            break;
+        }
+        case CacheLineState::SHARED:
+            bus->busReadExclusiveAndCheckIsExclusive(
+                this->getBlockIdx(tag),
+                threadID
+            );
+        case CacheLineState::FORWARD:
+            bus->setHasForwardState(false);
+            break;
+        case CacheLineState::EXCLUSIVE:
+        case CacheLineState::MODIFIED:
+            break;
+        default:
+            throw std::logic_error("Invalid cache state for MESIF");
+    }
+
+    logger->incrementPrivateDataAccess();
+    node->state = CacheLineState::MODIFIED;
+    return numCycles;
+}
+
+void MESIFCacheSet::handleBusReadEvent(
+    uint32_t tag,
+    std::shared_ptr<Logger> logger
+) {
+    if (!this->cacheSet.contains(tag)) return;
+
+    std::shared_ptr<CacheLineNode> node = this->cacheSet[tag];
+    
+    switch(node->state) {
+        case CacheLineState::FORWARD:
+        case CacheLineState::EXCLUSIVE:
+        case CacheLineState::MODIFIED:
+            logger->incrementBusTraffic();
+            logger->addExecutionCycles(this->numCyclesToSendBlock);
+            logger->addIdleCycles(this->numCyclesToSendBlock);
+        // In MESIF, cache-to-cache sharing never happens from SHARED state.
+        case CacheLineState::SHARED:
+            node->state = CacheLineState::SHARED;
+            break;
```

Comment:
nit (more explicit that shared state doesn't change):

```suggestion
        case CacheLineState::FORWARD:
        case CacheLineState::EXCLUSIVE:
        case CacheLineState::MODIFIED:
            logger->incrementBusTraffic();
            logger->addExecutionCycles(this->numCyclesToSendBlock);
            logger->addIdleCycles(this->numCyclesToSendBlock);
            node->state = CacheLineState::SHARED;
        // In MESIF, cache-to-cache sharing never happens from SHARED state.
        case CacheLineState::SHARED:
            break;
```